pkg/storage/util.go

```diff
@@ -106,23 +104,19 @@ func NamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
-		return "", fmt.Errorf("invalid name: %v", msgs)
-	}
-	return prefix + "/" + meta.GetNamespace() + "/" + name, nil
+	// We assume that NamespaceKeyFunc is used for read requests (List and Watch)
+	// and thus the object is correct and we don't perform any validation.
+	return prefix + "/" + meta.GetNamespace() + "/" + meta.GetName(), nil
 }
 
 func NoNamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	meta, err := meta.Accessor(obj)
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
```

Comment:
I would prefer we remove these, and use the keyfunc from the actual storage (it's just a coincidence that we can accurately calculate matching etcd keys using this separate function), and calculate the key once per object received, rather than calculating it per watcher